File: webviewerwidgets/wgnote/wgnote.py
import aiosqlite

async def get_data (queries):
    ret = {}
    try:
        dbpath = queries['dbpath']
        tab = queries['tab']
        rowkey = queries['rowkey']
        note = queries['note']
        if len(note) == 0:
            note = None
        conn = await aiosqlite.connect(dbpath)
        c = await conn.cursor()
        variant_cname = 'base__note_variant'
        gene_cname = 'base__note_gene'
        await c.execute('pragma table_info(variant)')
        variant_cols = [_[1] for _ in await c.fetchall()]
        if variant_cname not in variant_cols:
            variant_cname = 'base__note'
        await c.execute('pragma table_info(gene)')
        gene_cols = [_[1] for _ in await c.fetchall()]
        if gene_cname not in gene_cols:
            gene_cname = 'base__note'        
        if tab == 'variant':
            q = f'update {tab} set {variant_cname}=? where base__uid=?'
            await c.execute(q, (note, rowkey))
        elif tab == 'gene':
            q = f'update {tab} set {gene_cname}=? where base__hugo=?'
            await c.execute(q, (note, rowkey))
        await conn.commit()
        ret['status'] = 'success'
    except:
        import traceback
        traceback.print_exc()
        ret['status'] = 'fail'
    await c.close()
    await conn.close()
    return ret

File: webviewerwidgets/wgnote/wgnote.js
widgetGenerators['note'] = {
	'variant': {
		'width': 280, 
		'height': 80, 
		'donterase': true,
        'variables': {'noterow': ''},
		'function': function (div, row, tabName) {
            var widgetName = 'note';
			var v = widgetGenerators[widgetName][tabName]['variables'];
            if (v['noterow'] == row[0]) {
                return;
            } else {
                $(div).empty();
            }
            v['noterow'] = row[0];
            var colNo = infomgr.getColumnNo(tabName, 'base__note_variant');
            if (colNo === undefined) {
                colNo = infomgr.getColumnNo(tabName, 'base__note');
            }
            console.log(colNo)
            var note = row[colNo];
            var textbox = getEl('textarea');
            textbox.style.width = '100%';
            textbox.style.height = 'calc(100% - 24px)';
            textbox.value = note;
            addEl(div, textbox);
            addEl(div, getEl('br'));
            var button = getEl('button');
            button.textContent = 'Save';
            button.addEventListener('click', function (evt) {
                var note = evt.target.previousSibling.previousSibling.value;
                $grids[tabName].pqGrid('getData')[selectedRowNos[tabName]][colNo] = note;
                $grids[tabName].pqGrid('refresh');
                $.ajax({
                    url: '/result/runwidget/note',
                    data: {'job_id': jobId,
                           'dbpath': dbPath, 
                           'tab': tabName, 
                           'rowkey': row[0],
                           'note': note,
                          },
                    success: function (response) {
                    }
                });
            });
            addEl(div, button);
		}
	},
	'gene': {
		'width': 280,
		'height': 80,
		'donterase': true,
        'variables': {'noterow': ''},
		'function': function (div, row, tabName) {
            var widgetName = 'note';
			var v = widgetGenerators[widgetName][tabName]['variables'];
            if (v['noterow'] == row[0]) {
                return;
            } else {
                $(div).empty();
            }
            v['noterow'] = row[0];
            var colNo = infomgr.getColumnNo(tabName, 'base__note_gene');
            if (colNo === undefined) {
                colNo = infomgr.getColumnNo(tabName, 'base__note');
            }
            console.log(colNo)
            var note = row[colNo];
            var textbox = getEl('textarea');
            textbox.style.width = '100%';
            textbox.style.height = 'calc(100% - 24px)';
            textbox.value = note;
            addEl(div, textbox);
            addEl(div, getEl('br'));
            var button = getEl('button');
            button.textContent = 'Save';
            button.addEventListener('click', function (evt) {
                var note = evt.target.previousSibling.previousSibling.value;
                $grids[tabName].pqGrid('getData')[selectedRowNos[tabName]][colNo] = note;
                $grids[tabName].pqGrid('refresh');
                $.ajax({
                    url: '/result/runwidget/note',
                    data: {'job_id': jobId,
                           'dbpath': dbPath,
                           'tab': tabName, 
                           'rowkey': row[0],
                           'note': note,
                          },
                    success: function (response) {
                    }
                });
            });
            addEl(div, button);
		}
	}
}
